include/drivers/clock_control.h

```diff
@@ -166,24 +131,23 @@ static inline int clock_control_off(const struct device *dev,
 /**
  * @brief Request clock to start with notification when clock has been started.
  *
- * Function is non-blocking and can be called from any context.
- * When clock is already running user callback will be called from the context
- * of the function call else it is called from other context (e.g. clock
- * interrupt).
+ * Function is non-blocking and can be called from any context. User callback is
+ * called when clock is started.
  *
- * @param dev 	   Device.
- * @param sys	   A pointer to an opaque data representing the sub-system.
- * @param data	   Data structure containing a callback that is called when
- *		   action is performed. Structure content must be valid until
- *		   clock is started and user callback is called. Can be NULL.
+ * @param dev	    Device.
+ * @param sys	    A pointer to an opaque data representing the sub-system.
+ * @param cb	    Callback.
+ * @param user_data User context passed to the callback.
  *
- * @retval 0 if clock is started or already running.
- * @retval -EBUSY if same request already scheduled and not yet completed.
+ * @retval 0 if start is successfully initiated.
+ * @retval -EBUSY if clock was already started and is starting or running.
```

Comment:
This quite limits the usability of the function, as only the first requester will get the notification that clock has started.
Do we actually need to keep this function?

Reply:
yes, i think that as api does not define ref counting it should allow only single source which enables the subsystem

Reply:
Looking at the function description: https://github.com/zephyrproject-rtos/zephyr/blob/8316c653ecd51c7a9cd39ca64d39d4ff8f6d8042/include/drivers/clock_control.h#L132
I'm afraid such behavior may be surprising to some users. Perhaps some note stating this clearly could help?

And wouldn't `-EALREADY` be a more appropriate error code in this case?

Reply:
what shall we do with `clock_control_on`?
Now it does not specify what happens when called on already started clock. Ideally, it should probably return `-EALREADY` too but this change will impact more implementations so I would keep it without a change.

Reply:
IMO it should stay as it is. It should not matter to the requester if the clock has been started now or if it has been started earlier. This function is to just ensure that the clock is ready to be used.

Reply:
I don't really get the concept that zero can be the only success condition for API calls.  It's much more flexible to say non-negative is success, because you can convey additional information.

That said, I agree don't touch it the non-async functions in this PR as they weren't part of the change discussion.